llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp

```diff
@@ -885,53 +885,51 @@ template <typename Ty = llvm::Value> Ty *getVal(llvm::Metadata *M) {
+/// The value of the attribute is updated only if the new value \p NewVal is
+/// bigger than what is already stored in the attribute.
+static void updateGenXMDNodes(llvm::Function *F, genx::KernelMDOp MD,
+                              uint64_t NewVal) {
+  llvm::NamedMDNode *GenXKernelMD =
       F->getParent()->getNamedMetadata(GENX_KERNEL_METADATA);
-  assert(Nodes && "invalid genx.kernels metadata");
-  for (auto Node : Nodes->operands()) {
-    if (Node->getNumOperands() >= 4 && getVal(Node->getOperand(0)) == F)
-      return Node;
-  }
-  // if F is not a kernel, keep looking into its callers
-  while (!F->use_empty()) {
-    auto CI = cast<CallInst>(F->use_begin()->getUser());
-    auto UF = CI->getParent()->getParent();
-    if (auto Node = getSLMSizeMDNode(UF))
-      return Node;
+  assert(GenXKernelMD && "invalid genx.kernels metadata");
+  for (auto Node : GenXKernelMD->operands()) {
+    if (Node->getNumOperands() > MD &&
+        getVal(Node->getOperand(genx::KernelMDOp::FunctionRef)) == F) {
+
+      llvm::Value *Old = getVal(Node->getOperand(MD));
+      assert((Old && isa<llvm::ConstantInt>(Old)) &&
+             "integer constant expected");
+      uint64_t OldVal = cast<llvm::ConstantInt>(Old)->getZExtValue();
+      if (OldVal < NewVal) {
+        llvm::Value *New = llvm::ConstantInt::get(Old->getType(), NewVal);
+        Node->replaceOperandWith(MD, getMD(New));
+      }
+    }
   }
-  return nullptr;
-}
 
-static inline llvm::Metadata *getMD(llvm::Value *V) {
-  return llvm::ValueAsMetadata::get(V);
+  // Update all callers of 'F' as well.
+  for (auto It = F->use_begin(); It != F->use_end(); It++) {
+    auto CI = cast<CallInst>(It->getUser());
+    auto UserF = CI->getParent()->getParent();
+    updateGenXMDNodes(UserF, MD, NewVal);
+  }
 }
 
 static void translateSLMInit(CallInst &CI) {
   auto F = CI.getParent()->getParent();
-
   auto *ArgV = CI.getArgOperand(0);
   if (!isa<ConstantInt>(ArgV)) {
     assert(false && "integral constant expected for slm size");
     return;
   }
-  auto NewVal = cast<llvm::ConstantInt>(ArgV)->getZExtValue();
+  uint64_t NewVal = cast<llvm::ConstantInt>(ArgV)->getZExtValue();
   assert(NewVal != 0 && "zero slm bytes being requested");
-
-  // find the corresponding kernel metadata and set the SLM size.
-  if (llvm::MDNode *Node = getSLMSizeMDNode(F)) {
-    if (llvm::Value *OldSz = getVal(Node->getOperand(4))) {
-      assert(isa<llvm::ConstantInt>(OldSz) && "integer constant expected");
-      llvm::Value *NewSz = llvm::ConstantInt::get(OldSz->getType(), NewVal);
-      uint64_t OldVal = cast<llvm::ConstantInt>(OldSz)->getZExtValue();
-      if (OldVal < NewVal)
-        Node->replaceOperandWith(3, getMD(NewSz));
-    }
-  } else {
-    // We check whether this call is inside a kernel function.
-    assert(false && "slm_init shall be called by a kernel");
-  }
+  updateGenXMDNodes(F, genx::KernelMDOp::SLMSize, NewVal);
```

Comment:
Yes, if 'caller' need to reserve 1024 bytes inside it, but it also calls 'callee', which reserves 2048, then 'caller' need 2048 because of that call of 'callee'.
I did not change the logic here, just corrected it to avoid situation when 'caller' after inlining 'callee' reserves too small mem.
Such calls are normally (unless it is -O0) inlined by AlwaysInliner pass called after LowerESIMD.

Reply:
The scenario mentioned in my comment above is verified in the LIT test added to this PR.

For functions (not kernels) that potentially may be called from another module kernel , I added a TODO comment. I agree that such situations need to be specified and handled as well.